io/EptReader.cpp

```diff
@@ -902,86 +893,120 @@ Dimension::Type EptReader::getCoercedTypeTest(const NL::json& j)
-        startId = readLaszip(*m_bufferPointView, key, m_nodeId);
-    else
-        startId = readBinary(*m_bufferPointView, key, m_nodeId);
+    NodeBuffer(PointLayout& layout) : table(layout), view(table) { }
+    VectorPointTable table;
+    PointView view;
+};
 
-    log()->get(LogLevel::Debug) << "Points : "<< m_bufferPointView->size() <<
-        std::endl;
-    m_currentIndex = 0;
+void EptReader::load()
+{
+    // Asynchronously trigger the fetching and point-view execution of
+    // a lookahead buffer of nodes.
+    while (
+        m_upcomingNodeBuffers.size() < m_pool->size() &&
+        m_overlapIt != m_overlaps.end())
+    {
+        const auto nodeId(m_nodeId++);
+        const auto key(m_overlapIt->first);
+        ++m_overlapIt;
+
+        log()->get(LogLevel::Debug) << nodeId << "/" << m_overlaps.size() <<
+            std::endl;
+
+        // Insert an empty placeholder node to keep track of the outstanding
+        // nodes that are currently being fetched/executed.
+        std::unique_lock<std::mutex> lock(m_mutex);
+        auto& loadingBuffer = m_upcomingNodeBuffers[nodeId];
+        lock.unlock();
+
+        m_pool->add([this, &loadingBuffer, nodeId, key]()
+        {
+            std::unique_ptr<NodeBuffer> nodeBuffer(
+                new NodeBuffer(*m_userLayout));
 
-    // Read addon information after the native data, we'll possibly
-    // overwrite attributes.
-    for (const auto& addon : m_addons)
-        readAddon(*m_bufferPointView, key, *addon, startId);
+            if (m_info->dataType() == EptInfo::DataType::Laszip)
+                readLaszip(nodeBuffer->view, key, nodeId);
+            else
+                readBinary(nodeBuffer->view, key, nodeId);
 
-    m_nodeId++;
-}
+            for (const auto& addon : m_addons)
+                readAddon(nodeBuffer->view, key, *addon);
+
+            std::unique_lock<std::mutex> lock(m_mutex);
+            loadingBuffer = std::move(nodeBuffer);
+            lock.unlock();
 
+            // A node has been populated - notify our consumer thread.
+            m_cv.notify_one();
+        });
+    }
+}
 
-void EptReader::fillPoint(PointRef& point)
+EptReader::NodeBufferIt EptReader::findBuffer()
 {
-    DimTypeList dims = m_bufferPointView->dimTypes();
-    m_bufferPointView->getPackedPoint(dims, m_currentIndex,
-        m_temp_buffer.data());
-    point.setPackedData(dims, m_temp_buffer.data());
-    m_currentIndex++;
+    return std::find_if(
+        m_upcomingNodeBuffers.begin(),
+        m_upcomingNodeBuffers.end(),
+        [this](const NodeBufferPair& p) { return !!p.second; });
 }
 
+bool EptReader::next()
+{
+    // Asynchronously trigger the loading of some nodes.
+    load();
+
+    // Now wait for a node to be populated, or for there to be no outstanding
+    // nodes left, in which case we're all done.
+    std::unique_lock<std::mutex> lock(m_mutex);
+    m_cv.wait(lock, [this]()
+    {
+        return m_upcomingNodeBuffers.empty() ||
```

Comment:
Think this empty check isn't needed.

Reply:
How could it be omitted?  This structure being empty is our exit condition, and otherwise we will wait here forever once processing is complete and the `upcoming` list is empty.

Reply:
Isn't this wait the same as:

```
while (!upcomingNodeBuffers.empty() && findBuffer() == something)
   wait();
```
If the buffers are empty, you'll never get in the loop.  If they are, it's equivalent to saying that you found something.  In other words, you can't find something if the buffers are empty.

Reply:
I'm reading this: https://en.cppreference.com/w/cpp/thread/condition_variable/wait

Reply:
Maybe I'm just confused.  Replacing this check with simply `return findBuffer() != m_upcomingNodeBuffers.end();` ends up in an infinite `wait` at the end of processing like I'd expect - as you say, you can't find something if the buffers are empty.  And if the buffers are empty we want to exit for good, but without the `empty` check we hang because we're waiting for non-existent buffers to get populated.

What code exactly are you suggesting to be in this `wait` predicate?  I think I'm missing something.

Reply:
The confusion is mine.  I'm used to seeing it == container.end(), so that's what I was thinking.  Sorry.